src/test/java/com/microsoft/jenkins/azuread/integrations/casc/ConfigAsCodeTest.java

```diff
@@ -0,0 +1,87 @@
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+
+public class ConfigAsCodeTest {
+    private static final String TEST_UPN = "abc@jenkins.com";
+
+    @Rule
+    public JenkinsConfiguredWithCodeRule j = new JenkinsConfiguredWithCodeRule();
+
+    @Rule
+    public LoggerRule l = new LoggerRule().record(AzureAdMatrixAuthorizationStrategyConfigurator.class, Level.WARNING).capture(20);
+
+    @Test
+    @ConfiguredWithCode("configuration-as-code.yml")
+    public void should_support_configuration_as_code() {
+        SecurityRealm securityRealm = j.jenkins.getSecurityRealm();
+        assertTrue("security realm", securityRealm instanceof AzureSecurityRealm);
+        AzureSecurityRealm azureSecurityRealm = (AzureSecurityRealm) securityRealm;
+        assertEquals("clientId", azureSecurityRealm.getClientId());
+        assertEquals("clientSecret", azureSecurityRealm.getClientSecret());
+        assertEquals("tenantId", azureSecurityRealm.getTenant());
+
+        AuthorizationStrategy authorizationStrategy = j.jenkins.getAuthorizationStrategy();
+        assertTrue("authorization strategy", authorizationStrategy instanceof AzureAdMatrixAuthorizationStrategy);
+        AzureAdMatrixAuthorizationStrategy azureAdMatrixAuthorizationStrategy = (AzureAdMatrixAuthorizationStrategy) authorizationStrategy;
+
+        assertEquals("one real user sid", 2, azureAdMatrixAuthorizationStrategy.getAllSIDs().size());
+        assertTrue("anon can read", azureAdMatrixAuthorizationStrategy.hasExplicitPermission("anonymous", Jenkins.READ));
+        assertTrue("authenticated can read", azureAdMatrixAuthorizationStrategy.hasExplicitPermission(TEST_UPN, Jenkins.READ));
+        assertTrue("authenticated can build", azureAdMatrixAuthorizationStrategy.hasExplicitPermission("authenticated", Item.BUILD));
+        assertTrue("authenticated can delete jobs", azureAdMatrixAuthorizationStrategy.hasExplicitPermission(TEST_UPN, Item.DELETE));
+        assertTrue("authenticated can administer", azureAdMatrixAuthorizationStrategy.hasExplicitPermission(TEST_UPN, Jenkins.ADMINISTER));
+
+        assertEquals("no warnings", 0, l.getMessages().size());
+    }
+
+    @Test
+    @LocalData
+    public void export_configuration() throws Exception {
+        ConfiguratorRegistry registry = ConfiguratorRegistry.get();
+        ConfigurationContext context = new ConfigurationContext(registry);
+
+        SecurityRealm securityRealm = j.jenkins.getSecurityRealm();
+        Configurator realmConfigurator = context.lookupOrFail(AzureSecurityRealm.class);
+        @SuppressWarnings("unchecked")
+        CNode realmNode = realmConfigurator.describe(securityRealm, context);
+        assertNotNull(realmNode);
+        Mapping realMapping = realmNode.asMapping();
+        assertEquals(3, realMapping.size());
+        assertEquals("{clientId}", realMapping.getScalarValue("clientId"));
+
```

Comment:
please check that the client secret is encrypted here and not plain text

Reply:
any examples about this? I put the the encrypted credentials and it cannot decrypt correctly.

Reply:
pseudo code: 
`assert not equals "plain text client secret"`

`assert equals "plain text" ,Secret.decrypt(encrypted).getPlainText()`

Reply:
I use the local encrypted values in the test, but it cannot decrypt it in the test case. `Secret.decrypt` catch error `javax.crypto.BadPaddingException: Given final block not properly padded. Such issues can arise if a bad key is used during decryption.`.

Reply:
jenkinsci/credentials-plugin#114 also only has plain text for its password in test case, no idea how to mock encrypted one.

Reply:
use plain text it will encrypt it for you

Reply:
done